### DISCUSSION TITLE
ui.joystick

### DISCUSSION THREAD

**Discussion by bestec2** [zauberzeug/nicegui#299] [Q&A] [answered]
HI, what would be the easiest way to extract variables  x = and y = for coordinates. I mean something like x = f"{e.x:.3f}" .

```py
from nicegui import ui

ui.joystick(color='blue', size=50,
            on_move=lambda e: coordinates.set_text(f"{e.x:.3f}, {e.y:.3f}"),
            on_end=lambda _: coordinates.set_text('0, 0'))
coordinates = ui.label('0, 0')

ui.run()
```

**Comment by rodja** [comment, marked as answer]
If the values of an element should be used outside of the UI, you can store them in a model. This can be any Python object which the UI has access to. Here is an example using `dataclass`:

```python
from nicegui import ui

@dataclass
class Vector:
    x: float
    y: float

    def update(self, x, y):
        self.x = x
        self.y = y

v = Vector(0, 0)

ui.joystick(color='blue', size=50,
            on_move=lambda e: v.update(e.x, e.y),
            on_end=lambda _: v.update(0, 0))

coordinates = ui.label('0, 0')
ui.timer(0.1, lambda: coordinates.set_text(f'{v.x:.3f}, {v.y:.3f}'))

ui.run()
```

As you can see, storing the data into the Vector `v` is decoupled from displaying it in the `coordinates` label through a timer. 

Often you could also use the build-in [binding-mechanism](https://nicegui.io/reference#bindings). But then, the "value" you want to display in the label must be a single variable. I would probably do it like this:

```python
from nicegui import ui

@dataclass
class Data:
    direction: tuple[float, float] = (0, 0)

    def update(self, x, y):
        self.direction = (x, y)

data = Data()

ui.joystick(color='blue', size=50,
            on_move=lambda e: data.update(e.x, e.y),
            on_end=lambda _: data.update(0, 0))

ui.label('0, 0').bind_text_from(data, 'direction', backward=lambda d: f'{d[0]:.3f}, {d[1]:.3f}')

ui.run()
```

**Reply by falkoschindler** [reply]
@bestec2 This comment is somehow related: https://github.com/zauberzeug/nicegui/discussions/268#discussioncomment-4774052

**Reply by bestec2** [reply]
Thank you!
